server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java

```diff
@@ -107,30 +108,47 @@ public SerialTaskManagerWorker(Listener listener, Duration pollingInterval) {
 
     private Mono<Task.Result> run(TaskWithId taskWithId, Listener listener) {
         return Mono.from(listener.started(taskWithId.getId()))
-            .then(Mono.fromCallable(() -> runTask(taskWithId, listener)))
-            .onErrorResume(InterruptedException.class, e -> Mono.from(listener.cancelled(taskWithId.getId(), taskWithId.getTask().details())).thenReturn(Task.Result.PARTIAL))
+            .then(runTask(taskWithId, listener))
+            .onErrorResume(this::isCausedByInterruptedException, e -> cancelled(taskWithId, listener))
             .onErrorResume(Exception.class, e -> {
                 LOGGER.error("Error while running task {}", taskWithId.getId(), e);
                 return Mono.from(listener.failed(taskWithId.getId(), taskWithId.getTask().details(), e)).thenReturn(Task.Result.PARTIAL);
             });
     }
 
-    private Task.Result runTask(TaskWithId taskWithId, Listener listener) throws InterruptedException {
-        return taskWithId.getTask()
-            .run()
-            .onComplete(result -> Mono.from(listener.completed(taskWithId.getId(), result, taskWithId.getTask().details())).block())
-            .onFailure(() -> {
-                LOGGER.error("Task was partially performed. Check logs for more details. Taskid : " + taskWithId.getId());
-                Mono.from(listener.failed(taskWithId.getId(), taskWithId.getTask().details())).block();
-            });
+    private boolean isCausedByInterruptedException(Throwable e) {
+        if (e instanceof InterruptedException) {
+            return true;
+        }
+        return Optional.ofNullable(e.getCause())
+            .map(this::isCausedByInterruptedException)
```

Comment:
oh, a recursion, @blackheaven won't like it (:

What about 
```
        Stream.iterate(e, e.getCause() != null, e.getCause()).filter(e -> e instanceof InterruptedException).findAny();
```

Reply:
Arf, the call stack is expected to be small, I don't believe it makes much difference.